lib/smart_properties/property_collection.rb

```diff
@@ -9,28 +9,38 @@ def self.for(scope)
         ancestor.ancestors.include?(SmartProperties) && ancestor != SmartProperties
       end
 
-      parent.nil? ? new : new(parent.properties)
+      if parent.nil?
+        new
+      else
+        parent.properties.register(collection = new)
+        collection
+      end
     end
 
-    def initialize(parent = nil)
-      @parent = parent
+    def initialize
       @collection = {}
+      @collection_with_parent_collection = {}
+      @children = []
     end
 
     def []=(name, value)
+      name = name.to_s
       collection[name] = value
+      collection_with_parent_collection[name] = value
+      notify_children
+      value
     end
 
     def [](name)
-      collection_with_parent_collection[name]
+      collection_with_parent_collection[name.to_s]
     end
 
     def key?(name)
-      collection_with_parent_collection.key?(name)
+      collection_with_parent_collection.key?(name.to_s)
     end
 
     def keys
-      collection_with_parent_collection.keys
+      collection_with_parent_collection.keys.map(&:to_sym)
```

Comment:
Shouldn't this return strings? At the moment the method doesn't seem to be called anywhere, but following the GC argument, it would "ruin" everything with just one call…